src/ansi-c/c_nondet_symbol_factory.cpp

```diff
@@ -0,0 +1,403 @@
+    const typet &subtype=ns.follow(pointer_type.subtype());
+
+    code_labelt set_null_label;
+    code_labelt init_done_label;
+    std::string init_done_label_str;
+
+    if(!assume_non_null)
+    {
+      // Will generate the following code:
+      //   IF NONDET(_Bool) != FALSE THEN GOTO <set_null_label_str>
+      auto set_null_inst=
+        code_assignt(expr, null_pointer_exprt(pointer_type));
+      set_null_inst.add_source_location()=loc;
+
+      static size_t synthetic_constructor_count=0;
+      std::string set_null_label_str=
+        "post_synthetic_malloc_"+std::to_string(++synthetic_constructor_count);
+      set_null_label=code_labelt(set_null_label_str, set_null_inst);
+
+      init_done_label_str=set_null_label_str+"_init_done";
+      init_done_label=code_labelt(init_done_label_str, code_skipt());
+
+      code_ifthenelset null_check;
+      exprt null_return=from_integer(0, c_bool_typet(1));
+      null_check.cond()=
+        notequal_exprt(c_get_nondet_bool(c_bool_typet(1)), null_return);
+      null_check.then_case()=code_gotot(set_null_label_str);
+      init_code.move_to_operands(null_check);
+    }
+
+    exprt allocated=allocate_object(expr, subtype, loc, false);
+    exprt init_expr;
+    if(allocated.id()==ID_address_of)
+    {
+      init_expr=allocated.op0();
+    }
+    else
+    {
+      init_expr=dereference_exprt(allocated, allocated.type().subtype());
+    }
+    gen_nondet_init(init_expr, loc);
+
+    if(!assume_non_null)
+    {
+      // Will generate the following code:
+      //   GOTO <init_done_label_str>
+      //   <set_null_label_str>: <expr> = ((<pointer_type>)NULL);
+      //   <init_done_label_str>: SKIP
+      init_code.copy_to_operands(code_gotot(init_done_label_str));
+      init_code.move_to_operands(set_null_label);
+      init_code.move_to_operands(init_done_label);
+    }
+  }
+  else
+  {
+    side_effect_expr_nondett se=side_effect_expr_nondett(type);
+
+    // Will generate the following code:
+    //   <expr> = NONDET(type);
+    code_assignt assign(expr, se);
+    assign.add_source_location()=loc;
+    init_code.copy_to_operands(assign);
+  }
+}
+
+/*******************************************************************\
+
+Function: c_nondet_symbol_factory
+
+ Inputs:
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+exprt c_nondet_symbol_factory(
+  const code_typet::parametert &p,
+  const irep_idt base_name,
+  const typet &type,
+  code_blockt &init_code,
+  bool allow_null,
+  symbol_tablet &symbol_table,
+  const source_locationt &loc)
+{
+  input_recordert input_recorder(loc);
+  irep_idt identifier=id2string(goto_functionst::entry_point())+
+    "::"+id2string(base_name);
+
+  {
+    auxiliary_symbolt new_symbol;
+    new_symbol.mode=ID_C;
+    new_symbol.is_static_lifetime=false;
+    new_symbol.name=identifier;
+    new_symbol.base_name=base_name;
+    new_symbol.type=type;
+
+    symbol_table.move(new_symbol);
```

Comment:
Done. I also wrote the function comments for symbol_table.add() and symbol_table.move(), which I'll put in a separate PR.